L1Trigger/TrackerTFP/plugins/ProducerKFin.cc

```diff
@@ -0,0 +1,207 @@
+    StreamsTrack streamAcceptedTracks(dataFormats_->numStreams(Process::kf));
+    TTDTC::Streams streamLostStubs(dataFormats_->numStreams(Process::kf) * setup_->numLayers());
+    StreamsTrack streamLostTracks(dataFormats_->numStreams(Process::kf));
+    // read in SFout Product and produce KFin product
+    if (setup_->configurationSupported()) {
+      Handle<TTDTC::Streams> handleStubs;
+      iEvent.getByToken<TTDTC::Streams>(edGetTokenStubs_, handleStubs);
+      const TTDTC::Streams& streams = *handleStubs.product();
+      Handle<vector<TTTrack<Ref_Phase2TrackerDigi_>>> handleTTTracks;
+      iEvent.getByToken<vector<TTTrack<Ref_Phase2TrackerDigi_>>>(edGetTokenTTTracks_, handleTTTracks);
+      const vector<TTTrack<Ref_Phase2TrackerDigi_>>& ttTracks = *handleTTTracks.product();
+      for (int region = 0; region < setup_->numRegions(); region++) {
+        int nStubsSF(0);
+        for (int channel = 0; channel < dataFormats_->numChannel(Process::sf); channel++) {
+          const int index = region * dataFormats_->numChannel(Process::sf) + channel;
+          const TTDTC::Stream& stream = streams[index];
+          nStubsSF += accumulate(stream.begin(), stream.end(), 0, [](int& sum, const TTDTC::Frame& frame){ return sum += frame.first.isNonnull() ? 1 : 0; });
+        }
+        vector<StubSF> stubsSF;
+        stubsSF.reserve(nStubsSF);
+        for (int channel = 0; channel < dataFormats_->numChannel(Process::sf); channel++) {
+          const int index = region * dataFormats_->numChannel(Process::sf) + channel;
+          for (const TTDTC::Frame& frame : streams[index])
+            if (frame.first.isNonnull())
+              stubsSF.emplace_back(frame, dataFormats_);
+        }
+        vector<deque<TTDTC::Frame>> dequesStubs(dataFormats_->numChannel(Process::kf) * setup_->numLayers());
+        vector<deque<FrameTrack>> dequesTracks(dataFormats_->numChannel(Process::kf));
+        int i(0);
+        for (const TTTrack<Ref_Phase2TrackerDigi_>& ttTrack : ttTracks) {
+          if ((int)ttTrack.phiSector() / setup_->numSectorsPhi() != region) {
+            i++;
+            continue;
+          }
+          const int sectorPhi = ttTrack.phiSector() % setup_->numSectorsPhi();
+          deque<FrameTrack>& tracks = dequesTracks[sectorPhi];
+          const int binEta = ttTrack.etaSector();
+          const int binZT = dfZT.toUnsigned(dfZT.integer(ttTrack.z0()));
+          const int binCot = dfCot.toUnsigned(dfCot.integer(ttTrack.tanL()));
+          StubSF* stubSF = nullptr;
+          TTBV hitPattern(0, setup_->numLayers());
+          vector<int> layerCounts(setup_->numLayers(), 0);
+          for (const TTStubRef& ttStubRef : ttTrack.getStubRefs()) {
+            const int layerId = setup_->layerId(ttStubRef);
+            const int layerIdKF = layerEncoding_->layerIdKF(binEta, binZT, binCot, layerId);
+            hitPattern.set(layerIdKF);
```

Comment:
I thought TFP only needed TTTrack to interact with external code, or do truth association. But here, TTTrack is being used to derive the hit pattern. As the FW won't have access to TTTrack, but only to the info in stubsSF, wouldn't it be better to derive it from the latter?

Reply:
TTTrack is not being used to derive the hit pattern, the hit pattern field in TTrack is used to transport the channel id. The channel id is known to the f/w.